src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp

```diff
@@ -20,8 +21,14 @@ void shape_inference(ov::Node* op,
 
 class IShapeInferCommon {
 public:
-    virtual std::vector<StaticShape> infer(const std::vector<StaticShape>& input_shapes,
-                                           const std::map<size_t, HostTensorPtr>& constant_data) = 0;
+    struct ShapeInferResult {
+        std::vector<StaticShape> shapes;
+        ShapeInferStatus status;
+    };
+
+public:
+    virtual ShapeInferResult infer(const std::vector<StaticShape>& input_shapes,
```

Comment:
As the infer result status has got only two states now (not sure if there is a plan to expand it), maybe there is a chance to use empty vector as skipped status?

In template implementation of shape infer it looks like it always returns not empty vector or throw exception if inference fails. The empty  vector could be interpreted as infer has been skipped. Then return type of this function doesn't need to be changed, and changes will be less impacting especially on classes where this status has no special meaning and they always return `updated`.

Other solution (to expand interface not modify it) can be add additional function to this interface e.g. `get_status()` which provide last infer status, for most classes it can be compile time value as it never changes.
Then `entryBase` can implement this getter with default (updated) value and only in required classes override it.

Reply:
Thank you for your suggestions.

>In template implementation of shape infer it looks like it always returns not empty vector or throw exception if inference fails. The empty vector could be interpreted as infer has been skipped. Then return type of this function doesn't need to be changed, and changes will be less impacting especially on classes where this status has no special meaning and they always return `updated`.

Of course, this is the first idea that came to my mind. But, what if the empty vector is a result of some error? In that case the program can not distinguish whether the empty vector is `skip` or we got unexpectedly empty data. Thus, we will have hardly detecting error. The sing is on the `IShapeInferCommon` hierarchy level we do not have control over the template `shape_infer` functions, and in the case of using an empty vector as a `skip` indication, we may just "misrepresent" an error as an valid result. But with the approach proposed in the PR we clearly indicate what we expect. Also, since this "complexity" is located on this level is think this is not a big problem.

>Other solution (to expand interface not modify it) can be add additional function to this interface e.g. get_status() which provide last infer status, for most classes it can be compile time value as it never changes.
Then `entryBase` can implement this getter with default (updated) value and only in required classes override it.

In fact, this approach does not reduce the complexity, but makes the interface less convenient, since the user should remember about the state that should be read with `get_status()`. Yes, in this proposal the user will also have to check the return status, but he receive it together with the resulting shape, so this is less likely that he forget to check it, so slightly more convenient.

But, to be frank, I do not completely like the solution either, but given the options we have, I think the proposed solution is better than others. Also in the CPU plugin I introduced the same `Result` type and it can be used to optimize the shape infer call. We can simply skip the computations if the input data have not been changed from the previous call.
If my reasoning is not good enough, further discussion is always welcome (:

Reply:
Change interface is also fine, just was thinking about other solutions.
For me it looks like the `std::optional` from c++17 could be useful here.

Reply:
Completely agree with you. But we cannot use c++17 so far.